sdk/communication/Azure.Communication.ProgrammableConnectivity/src/Generated/NumberVerification.cs

```diff
@@ -0,0 +1,348 @@
+        {
+            Argument.AssertNotNull(apcGatewayId, nameof(apcGatewayId));
+            Argument.AssertNotNull(numberVerificationWithoutCodeContent, nameof(numberVerificationWithoutCodeContent));
+
+            RequestContext context = FromCancellationToken(cancellationToken);
+            using RequestContent content = numberVerificationWithoutCodeContent.ToRequestContent();
+            Response response = VerifyWithoutCode(apcGatewayId, content, context);
+            return response;
+        }
+
+        /// <summary>
+        /// [Protocol Method] Verifies the phone number (MSISDN) associated with a device. As part of the frontend authorization flow, the device is redirected to the operator network to authenticate directly.
+        /// <list type="bullet">
+        /// <item>
+        /// <description>
+        /// This <see href="https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/ProtocolMethods.md">protocol method</see> allows explicit creation of the request and processing of the response for advanced scenarios.
+        /// </description>
+        /// </item>
+        /// <item>
+        /// <description>
+        /// Please try the simpler <see cref="VerifyWithoutCodeAsync(string,NumberVerificationWithoutCodeContent,CancellationToken)"/> convenience overload with strongly typed models first.
+        /// </description>
+        /// </item>
+        /// </list>
+        /// </summary>
+        /// <param name="apcGatewayId"> The identifier of the APC Gateway resource which should handle this request. </param>
+        /// <param name="content"> The content to send as the body of the request. </param>
+        /// <param name="context"> The request context, which can override default behaviors of the client pipeline on a per-call basis. </param>
+        /// <exception cref="ArgumentNullException"> <paramref name="apcGatewayId"/> or <paramref name="content"/> is null. </exception>
+        /// <exception cref="RequestFailedException"> Service returned a non-success status code. </exception>
+        /// <returns> The response returned from the service. </returns>
+        /// <include file="Docs/NumberVerification.xml" path="doc/members/member[@name='VerifyWithoutCodeAsync(string,RequestContent,RequestContext)']/*" />
+        public virtual async Task<Response> VerifyWithoutCodeAsync(string apcGatewayId, RequestContent content, RequestContext context = null)
+        {
+            Argument.AssertNotNull(apcGatewayId, nameof(apcGatewayId));
+            Argument.AssertNotNull(content, nameof(content));
+
+            using var scope = ClientDiagnostics.CreateScope("NumberVerification.VerifyWithoutCode");
+            scope.Start();
+            try
+            {
+                using HttpMessage message = CreateVerifyWithoutCodeRequest(apcGatewayId, content, context);
+                RedirectPolicy.SetAllowAutoRedirect(message, true);
```

Comment:
@jsquire we don't want the 302 to be followed, we just want it to be returned, for the developer using the SDK to extract the `location` header, and then follow it themselves.

See our typespec: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/number.tsp

How can I set this to `RedirectPolicy.SetAllowAutoRedirect(message, false);`? I'm assuming that manually changing SDK code is not allowed

Reply:
I don't have an answer to this one.   @m-nash: Can you offer guidance?

Reply:
When manually changing the generated SDK code, my test successfully passes in live mode. I will definitely need to either change that manually or have another solution before releasing.

Reply:
You cannot merge with generated code that has been hand-edited.  There's a gate that will fail CI.

Reply:
@annelo-msft is there a way to add an internal policy so that 302 is not followed as redirect without modifying the generated functions themselves?

Reply:
Can you clarify what you're trying to accomplish?  Is it that you want to follow the redirect from this endpoint for all cases _except_ 302?  Or do you never want to follow the redirect for this endpoint?

Reply:
@annelo-msft Our case is that for calls to the `number-verification` endpoint, we're EXPECTING a 302 back as a response. This is described in our typespec. Therefore, we want to pass that 302 back to the developer using our SDK, so they can follow the `location` header. We can get exactly this behaviour if I change: `RedirectPolicy.SetAllowAutoRedirect(message, false);`

We need to do this because the URL in the `location` header can only be followed on specific networks, which won't be the network that the SDK is called from.

Reply:
I see.  It sounds like you are saying that in no case do you want to follow the redirect.  Is that correct?

If you never want to follow the redirect, and you want to return to the caller on a 302, then that should be the default behavior that you get if the `RedirectPolicy.SetAllowAutoRedirect` line is removed entirely.

I don't know offhand what in your TypeSpec file triggers the generator to add that line, but maybe @m-nash can explain if he understands that part?  I would expect that modifying the TypeSpec such that that line is removed should achieve your goals, but if that doesn't do it and you need further input regarding how Azure.Core types work, I'm happy to continue to help here.

Thanks!

Reply:
@annelo-msft  Yes, that is correct, we never want to follow the redirect. Can I get some guidance towards what I should be changing in my Typespec please? https://github.com/Azure/azure-rest-api-specs/blob/main/specification/programmableconnectivity/Azure.ProgrammableConnectivity/apis/number.tsp here's the typespec for `number-verification` specifically, where you can see 

```
interface NumberVerification {
  #suppress "@azure-tools/typespec-azure-core/no-response-body" "302 redirect has no response body"
  @doc("Verifies the phone number (MSISDN) associated with a device. As part of the frontend authorization flow, the device is redirected to the operator network to authenticate directly.")
  @sharedRoute
  @action("verify")
  verifyWithoutCode is Operations.ResourceAction<
    NumberVerificationEndpoint,
    NumberVerificationWithoutCodeContent,
    TypeSpec.Http.Response<302> & {},
    TraitOverride<ResponseHeadersTrait<{
      @doc("The URI of the network's authorization endpoint, which should be followed by the front-end application.")
      @header
      location: url;
    } & RequestIdResponseHeader>>
  >;
  ```
  
The 302 is defined. It's a relatively fast repo to change the typespec, and regenerate the SDK, so I'm happy to test any changes recommended. I'd love to be able to resolve this ASAP, because we've agreed with some developers that we'll have an SDK out by early May.

Reply:
@annelo-msft @m-nash bumping this thread as it's a blocker for skipping our SDK. Could you please advise on next steps? Happy to jump on a call and discuss, and open to changing typespec etc, but I'd need to make sure that it gets discussed with the archboard, which takes time as well.

The only other thing I see is just manually editing the SDK generated code, but as I understand that, it'd fail some pipeline. But if there's a way to override that, then I'd be happy with that solution too.

Reply:
@AndrewNivenMSFT tagging as backup owner of SDKs for our team

Reply:
This appears to be a bug in the generator logic, where if a service defines a redirect response code, the client is generated with a policy that opts-into following it.    The intention is that clients do not follow redirects by default and must opt-in to do so.   We'll work on getting this fixed, but in the short term, I see two options:

- Remove the 302 response from your TypeSpec, which will cause the generator to not add the redirect policy to the client.   (@m-nash:  Please confirm that I'm not missing a step)
- Add a customization to the client which either removes the policy or opts-out.   (@annelo-msft:  Do we have any examples of this that we can share?  I think [Monitor](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/ApplicationInsightsRestClient.cs) would be one, no?)

Reply:
> will cause the generator to not add the redirect policy to the client.

To be clear, the redirect policy is always in the client's pipeline.  It's just that this line of code sets a property on the _message_ that [tells the redirect policy it's ok to follow the redirect](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Pipeline/RedirectPolicy.cs#L59) when the response flows through it.

So, if you want to disable redirects for the whole client, you can add a policy to your pipeline that resets the message property to false.  Such a policy would look like this:

```csharp
class DontFollowRedirectPolicy : HttpPipelinePolicy
{
    public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
    {
        RedirectPolicy.SetAllowAutoRedirect(message, false);

        ProcessNext(message, pipeline);
    }

    public override async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
    {
        RedirectPolicy.SetAllowAutoRedirect(message, false);

        await ProcessNextAsync(message, pipeline);
    }
}
```

You'll want to add this as a per-call policy so it comes before the redirect policy in the pipeline.

It looks like the line of code that creates the pipeline for this subclient is here: https://github.com/Azure/azure-sdk-for-net/blob/b919fcc8057a3091bce2afb1d3f30ee430d23d19/sdk/communication/Azure.Communication.ProgrammableConnectivity/src/Generated/ProgrammableConnectivityClient.cs#L55

This would change to something like this:
```diff
- _pipeline = HttpPipelineBuilder.Build(options, Array.Empty<HttpPipelinePolicy>(), new HttpPipelinePolicy[] { new BearerTokenAuthenticationPolicy(_tokenCredential, AuthorizationScopes) }, new ResponseClassifier());
+ _pipeline = HttpPipelineBuilder.Build(options, new HttpPipelinePolicy[] { new DontFollowRedirectPolicy() }, new HttpPipelinePolicy[] { new BearerTokenAuthenticationPolicy(_tokenCredential, AuthorizationScopes) }, new ResponseClassifier());
```

Reply:
@annelo-msft thank you for the quick response! I can see how your proposed code change would work. However, since it's in `src/Generated`, doesn't that mean that I can't change that bit of code? Otherwise I could just delete the 2 lines of code I proposed earlier.

Reply:
@annelo-msft 

I'm afraid I don't follow your proposed solution. Could you please give a bit more concrete code/procedure of how you think it'll work?
 
I can't change any of the code in `communication\Azure.Communication.ProgrammableConnectivity\src\Generated\ProgrammableConnectivityClient.cs`
, because after I run `dotnet build /t:GenerateCode`, it just gets reverted. And in my test code, an example would be:
 
 ```
string apcGatewayId = "/subscriptions/abcdefgh/resourceGroups/.../Microsoft.programmableconnectivity/...";
Uri endpoint = new Uri("https://your-endpoint-here.com");
TokenCredential credential = new DefaultAzureCredential();
ProgrammableConnectivityClient baseClient = new ProgrammableConnectivityClient(endpoint, credential);
DeviceLocation client = baseClient.GetDeviceLocationClient();
DeviceLocationVerificationContent content = new DeviceLocationVerificationContent(
    networkIdentifier: new NetworkIdentifier("NetworkCode", "Telefonica_Brazil"),
    latitude: 80.0,
    longitude: 85.1,
    accuracy: 50,
    device: new LocationDevice
    {
        PhoneNumber = "+8000000000000"
    }
);
Response<DeviceLocationVerificationResult> result = await client.VerifyAsync(apcGatewayId, content);
 ```
 
I don't see how I can insert a custom pipeline into that code, when constructing `ProgrammableConnectivityClient` etc. Could you please clarify?

Reply:
@m-nash, would you be able to share links to generator documentation describing how to do this?

> you can customize the constructor by adding a partial type that provides a constructor with the same signature as the generated constructor. When you re-generate, the generator will no longer write this constructor.

@aronhegedusms, while we're waiting for the official docs, if you have a look at https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/communication/Azure.Communication.CallAutomation/src in the ACS solution - you'll see that there is a top-level client defined in that directory that adds constructors.  This, I believe, is the "customization layer."  The generator uses Roslyn APIs to see what's been customized and doesn't add methods/constructors with existing signatures in the "generated layer."  So, you should be able to add a custom client with the constructor signature of your generated client and add the lines to add the custom policy into the pipeline when it is created.

Let us know if you have further questions about this, and thanks!

Reply:
@lirenhe:  Would you be able to assist with a pointer to generator documentation?

Reply:
have fixed this as suggested. Thank you